src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/AbstractPersistentStorage.cs

```diff
@@ -34,16 +39,42 @@ internal abstract class AbstractPersistentStorage : IChecksummedPersistentStorag
         public abstract void Dispose();
 
         public abstract Task<Checksum> ReadChecksumAsync(string name, CancellationToken cancellationToken);
-        public abstract Task<Checksum> ReadChecksumAsync(Project project, string name, CancellationToken cancellationToken);
-        public abstract Task<Checksum> ReadChecksumAsync(Document document, string name, CancellationToken cancellationToken);
 
-        public abstract Task<Stream> ReadStreamAsync(string name, Checksum checksum, CancellationToken cancellationToken);
-        public abstract Task<Stream> ReadStreamAsync(Project project, string name, Checksum checksum, CancellationToken cancellationToken);
-        public abstract Task<Stream> ReadStreamAsync(Document document, string name, Checksum checksum, CancellationToken cancellationToken);
+        protected abstract Task<Checksum> ReadChecksumAsync(ProjectKey projectKey, Project? projectOpt, string name, CancellationToken cancellationToken);
+        protected abstract Task<Checksum> ReadChecksumAsync(DocumentKey documentKey, Document? documentOpt, string name, CancellationToken cancellationToken);
+
+        public Task<Checksum> ReadChecksumAsync(ProjectKey projectKey, string name, CancellationToken cancellationToken)
+            => ReadChecksumAsync(projectKey, projectOpt: null, name, cancellationToken);
+
+        public Task<Checksum> ReadChecksumAsync(DocumentKey documentKey, string name, CancellationToken cancellationToken)
+            => ReadChecksumAsync(documentKey, documentOpt: null, name, cancellationToken);
```

Comment:
when passing the 'key' form, we jsut call the main method with 'null' passed for the snapshot.